# Mạng nơ-ron với scikit-learn

Mục tiêu của notebook này là xây dựng mô hình DNN sử dụng thư viện scikit-learn để dự đoán cảm xúc từ các bài bình luận sản phẩm. Chúng ta sẽ thực hiện:

 * Load các bình luận về sản phẩm.
 * Triển khai mô hình DNN với scikit-learn.
 * Điều chỉnh một vài tham số.

In [1]:
# Import một số thư viện

import pandas
import numpy as np
from utils import get_product_reviews_data

## Load tập dữ liệu product reviews

Như mô-đun trước, chúng ta sẽ load, tiền xử lý dữ liệu, chuyển đổi và chia chúng thành các tập huấn luyện và kiểm tra. Trong notebook này, chúng ta không tập trung vào điều đó nên chỉ cần chạy các cell sau. Có thể kiểm tra code dữ liệu bên trong folder **utils**.

In [2]:
train_set, val_set = get_product_reviews_data()

sentiment_X_train, sentiment_y_train = train_set
sentiment_X_valid, sentiment_y_valid = val_set

/Users/ngothai/opt/miniconda3/envs/tf/lib/python3.7/site-packages/pandas/core/indexing.py:1667: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  self.obj[key] = value


# Xây dựng phân loại DNN với scikit-learn
Giờ hãy sử dụng DNN có sẵn của sklearn: [sklearn.neural_network.MLPClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html#sklearn.neural_network.MLPClassifier).

## Chọn kích thước lớp ẩn
Trước tiên, hãy kiểm tra số lượng đặc trưng ở từng tập dữ liệu rồi chọn kích thước lớp ẩn thích hợp.

In [ ]:
print ("Sentiment feature size: ", sentiment_X_train.shape[1])

Sentiment feature size:  194


Với tập dữ liệu sentiment, chúng ta sẽ chọn 2 lớp ẩn có 64, 8 nơ-ron tương ứng.

In [ ]:
from sklearn.neural_network import MLPClassifier

# # number of hidden layers and number of neurons in each layer is 64, 8
clf_sentiment = MLPClassifier(solver='adam', alpha=1e-5, activation='relu',
                    hidden_layer_sizes=(64, 8), random_state=1) 

clf_sentiment.fit(sentiment_X_train, sentiment_y_train)

print ("***Sentiment result***")
print("Train accuracy: {}".format(clf_sentiment.score(sentiment_X_train, sentiment_y_train)))
print("Validation accuracy: {}".format(clf_sentiment.score(sentiment_X_valid, sentiment_y_valid)))

***Sentiment result***
Train accuracy: 0.9614206981016534
Validation accuracy: 0.7157527793480309


/Users/ngohongthai/Documents/projects/tensorflow-env/env/lib/python3.8/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Như các bạn thấy, DNN đơn giản có các kết quả khá hợp lý và mạng nơ-ron có thể khớp khá tốt với mẫu tronng tập huấn luyện.
<br>
**Quiz**: Validation accuracy là bao nhiêu?
<br>
**Đáp án**: 0.7157527793480309